core/include/detray/tools/intersection_kernel.hpp

```diff
@@ -21,179 +21,92 @@ namespace detray {
-        std::tuple_size_v<typename mask_container::mask_tuple> - 1>(
-        intersection, links, track, ctf, masks, range, mask_context);
+
+    // No intersection was found
+    return intersection{};
 }
 
-/** Actual kernel method that updates the updates the intersections
+/** Kernel method that updates the intersections
  *
- * @tparam surface_type is The type of the surface container
- * @tparam transform_container is the type of the transform container
- * @tparam mask_container is the type of the type of the mask container
- * @tparam link_type is the type of the geometry links, that belong to the
- *         surface type
+ * @tparam track_type The type of the track/context
+ * @tparam surface_type The type of the surface container
+ * @tparam transform_container The type of the transform container
+ * @tparam mask_container The type of the type of the mask container
  *
  * @param track the track information including the contexts
  * @param surface the surface type to be intersected
  * @param contextual_transform the transform container
  * @param masks the tuple mask container to for the intersection
- * @param mask_links the geometry links discovered by the intersection
  *
- * @return an intersection and the link to the result
+ * @return  an intersection struct (invalid if no intersection was found)
  **/
 template <typename track_type, typename surface_type,
-          typename transform_container, typename mask_container,
-          typename link_type>
-const auto intersect(const track_type &track, surface_type &surface,
-                     const transform_container &contextual_transforms,
-                     const mask_container &masks, link_type &mask_links) {
+          typename transform_container, typename mask_container>
+inline const auto intersect(const track_type &track, surface_type &surface,
+                            const transform_container &contextual_transforms,
+                            const mask_container &masks) {
+    // Gather all information to perform intersections
     const auto &ctf = contextual_transforms[surface.transform()];
-    auto mask_link = surface.mask();
-    const auto &mask_context = std::get<0>(mask_link);
+    const auto &volume_index = surface.volume();
+    const auto &mask_link = surface.mask();
+    const auto &mask_id = std::get<0>(mask_link);
     const auto &mask_range = std::get<1>(mask_link);
 
-    // Create a return intersection and run the variadic unrolling
-    intersection result_intersection;
-
     // Unroll the intersection depending on the mask container size
-    unroll_intersect(
-        result_intersection, mask_links, track, ctf, masks, mask_range,
-        mask_context,
+    return unroll_intersect(
+        track, ctf, masks, mask_range, mask_id, volume_index,
         std::make_integer_sequence<
-            dindex, std::tuple_size_v<typename mask_container::mask_tuple>>{});
```

Comment:
I think this change is not very necessary

Reply:
Yes, that was the tricky part. What I wanted to try, was to use this both with ```mask_store``` and raw tuple types

Reply:
OK